### PR TITLE
Compare libdependency and alternative implementations

### PR DESCRIPTION
**Pull request by ryanmoran**
## Summary

⚠️ DO NOT MERGE ⚠️ 

There are 2 implementations in this PR that compare what the implementation looks like using `libdependency` and a significantly trimmed down alternative.

The code implementations can be found in [`dependency/retrieval/main.go`](https://github.com/paketo-buildpacks/go-dist/blob/dependencies/dependency/retrieval/main.go). Shared code between the implementations exists in [`dependency/retrieval/releases.go`](https://github.com/paketo-buildpacks/go-dist/blob/dependencies/dependency/retrieval/releases.go). And an alternative implementation can be found in [`dependency/retrieval/alternative`](https://github.com/paketo-buildpacks/go-dist/blob/dependencies/dependency/retrieval/alternative).

### CODE REVIEW

File: dependency/retrieval/main.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		var versions []string
+		for _, release := range releases {
+			versions = append(versions, release.SemVer.String())
+		}
+
+		return versionology.NewSimpleVersionFetcherArray(versions...)
+	}
+
+	generateMetadataFunc := func(version versionology.VersionFetcher) (versionology.Dependency, error) {
+		release, err := fetcher.Get(version.Version().String())
```
Comment (sophiewigmore) [inline dependency/retrieval/main.go:32]: To me, the types here would be the hardest thing to digest when using `libdependency`.
Specifically around `version.Version().String()`, having to run the `Version()` function on a variable already named `version`, and then converting to a string is confusing. (I get the code in libdependency, I'm just noting that from looking at the implementation it's not the most intuitive)

**Comment by sophiewigmore** [comment]
Did you find using `retrieve.NewMetadata` / creating `versionFetcherFunc` and `generateMetadataFunc` intuitive? Were the docs helpful?

What was your primary motivation for creating an alternative implementation?

**Comment by sophiewigmore** [comment]
cc @joshuatcasey if you are interested

**Comment by ryanmoran** [comment]
@sophiewigmore I think the biggest issue for me was that, as a library, it felt kind of "closed" to me. I didn't feel like I had a set of pieces to build something with. It felt like I had a tightly defined structure I had to fit within.

I wrote down some notes while I was putting together the `libdependency`-based implementation. They're mostly stream of consciousness, so let me know if anything needs more context. Let me post them here:

* `retrieve.NewMetadata`
  * My immediate thought was that this was a constructor (like `NewSomething`, etc.), but it doesn't return anything. I think I can get behind it being a high-level orchestrator that creates the `metadata.json` file as a side effect, but it should at the least return any errors it encounters, no?
  * It panics on errors. That doesn't really seem like an ideal API. What if I want to clean something up before the program exits, am I expected to `recover`?
  * The biggest mental barrier to understanding this is to recognize that its a function implementing inversion of control where I am required to supply some custom steps in the predefined process of generating the `metadata.json` file. The alternative I supplied in this PR just does away with that inversion and leaves it up to the user to call into the library imperatively.
* `retrieve.FetchArgs`:
  * Why can't I parse these args myself? The first thing I thought to do with my `retrieval` program was to pass the 2 flags I invoke for this command into something. But there's nothing to pass them into, they get read magically from within the `NewMetadata` function. If that's the desired purpose, then why is the helper function exported?
* `libdependency.VersionFetcherFunc`:
  * What is this? I need to implement it for `retrieve.NewMetadata`, but there's no documentation to explain what its purpose is. I eventually figured out that its a function that returns a list of all known versions for a dependency.
  * Why is it defined at the top-level and not within the `retrieve` package? Is it used somewhere else?
* `versionology.VersionFetcher`:
  * This is a confusing set of type names when paired with the `libdependency.VersionFetcherFunc`. Arguably the `VersionFetcherFunc` is a `VersionFetcher` and the `VersionFetcher` is a `Version`. I find that I've got `version.Version().String()` in a couple of places and that feels odd.
  * I still don't fully grasp why the `versionology` package exists. In my alternative implementation, I pass around strings with the contract that they be parsable SemVer versions.
* `retrieve.GeneratePURL` and `retrieve.LookupLicenses`:
  * I understand why these helpers exist. I don't understand why they are in the `retrieve` package. Their purpose just feels orthogonal to the primary purpose of the `retrieve` package.
* `workflows.ToWorkflowJson`:
  * Could this be inlined into `retrieve.NewMetadata`? Its only ever called from that one location.

**Comment by joshuatcasey** [comment]
@ryanmoran keep in mind that the PR into `libdependency` (https://github.com/paketo-buildpacks/libdependency/pull/1) is coming from a different repo than what's used here. The docs in that PR (https://github.com/joshuatcasey/libdependency-fork/tree/pr-to-upstream) are definitely more helpful than the docs in the POC repo (https://github.com/joshuatcasey/libdependency).

Unfortunately because Github is strict on the naming conventions you can't import from the fork :(. Once we get that PR merged it will have the right name (github.com/paketo-buildpacks/libdependency).

**Comment by joshuatcasey** [comment]
- The name `NewMetadata` is meant to be used with the package name, so it reads `retrieve.NewMetadata`. I get that's a little obtuse but it's at least inspired by https://go.dev/doc/effective_go#names. I get that this conflicts a bit with memory allocation but that could be easily fixed by calling it `retrieve.UnseenMetadata` or something. 
- `retrieve.FetchArgs` is only public for testing purposes, the `godoc` should call that out. It's apparently strangely difficult to test code that relies on `flag`.
- `libdependency.VersionFetcherFunc`: I've moved this to `retrieve` and renamed it so now it's `retrieve.GetAllVersionsFunc`. Hopefully this helps. Also added `godoc`.
- `retrieve.GeneratePURL` and `retrieve.LookupLicenses`: They're part of retrieving metadata and the assumption is that buildpack authors will consume them from within their implementation of `generateMetadata`. 
- `workflows.ToWorkflowJson`: sure I'll inline if that helps clarify what's being used where. I suspect it will reappear but that's a problem for another time.

As for panic'ing, that's explicitly what I want out of `libdependency`. It's an internal *and optional* tool for buildpack authors to use primarily within workflows. If a panic occurs, the workflow fails and the buildpack author can debug via output, it's not intended to allow for recovery or cleanup.

**Pull request closed without merging**